emission/tests/storageTests/TestTimeSeries.py

```diff
@@ -81,6 +81,76 @@ def testExtraQueries(self):
         with self.assertRaises(AttributeError):
             list(ts.find_entries(time_query=tq, extra_query_list=[ignored_phones]))
 
+    def testFindEntriesCount(self):
+        '''
+        Test: Specific keys with other parameters not passed values.
+        Input: For each dataset: ["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+            - Testing this with sample dataset: "shankari_2015-aug-21", "shankari_2015-aug-27"
+        Output: Aug_21: ([738, 508], [0]), Aug_27: ([555, 327], [0])
+            - Actual output just returns a single number for count of entries.
+            - Validated using grep count of occurrences for keys: 1) "background/location"     2) "background/filtered_location"
+                - $ grep -c <key> <dataset>.json
+        
+        For Aggregate Timeseries test case:
+        - The expected output would be summed-up values for the respective keys from the individual users testing outputs mentioned above.
+        - Output: ([1293, 835], [0])
+            - For each of the 3 input keys from key_list1: 
+                - 1293 = 738 (UUID1) + 555 (UUID2)
+                - 835 = 508 (UUID1) + 327 (UUID2)
+                - 0 = 0 (UUID1) + 0 (UUID2)
+
+        '''
+
+        ts1_aug_21 = esta.TimeSeries.get_time_series(self.testUUID1)
+        ts2_aug_27 = esta.TimeSeries.get_time_series(self.testUUID)
+
+        # Test case: Combination of original and analysis timeseries DB keys for Aug-21 dataset
+        key_list1=["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+        count_ts1 = ts1_aug_21.find_entries_count(key_list=key_list1)
+        self.assertEqual(count_ts1, ([738, 508], [0]))
+
+        # Test case: Combination of original and analysis timeseries DB keys for Aug-27 dataset
+        key_list1=["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+        count_ts2 = ts2_aug_27.find_entries_count(key_list=key_list1)
+        self.assertEqual(count_ts2, ([555, 327], [0]))
+
+        # Test case: Only original timeseries DB keys for Aug-27 dataset
+        key_list2=["background/location", "background/filtered_location"]
+        count_ts3 = ts2_aug_27.find_entries_count(key_list=key_list2)
+        self.assertEqual(count_ts3, ([555, 327], []))
```

Comment:
And here it's an empty array. Why the inconsistency?

Reply:
Empty array is returned in case there were no keys pertaining to the respective timeseries database.
In this specific test case, only background keys are there, hence their count is returned.
No, analysis keys are present in the input keys, hence empty array returned.

This is to differentiate from the [0] case where a key might be present in the input but no matching documents found.
Whereas in this case of [], no key was present in the input itself.

Reply:
That is an interesting design choice, but I am not sure I agree with it. This seems to:
1. **very tied to this specific implementation:** e.g. the difference between timeseries and analysis_timeseries. Whenever you are designing an abstract interface, think about what it would look like if the underlying implementation were completely different. What if we stored everything in giant csv files instead? What if we used a real timeseries database instead? What would the meaning of `timeseries` and `analysis_timeseries` be in that case?
2. **unnecesarily complex to use:** again, think of the user of the interface, who is using the timeseries abstraction that we have outlined in chapter 5 of my thesis. they want to know how many entries there are in each stream. Do they want to know where the entries came from? Do they want to know the details of our implementation? Do they want to know the difference between the key being present but there being zero entries and the key being absent? No. We shouldn't make them do that. Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries.

Reply:
Thank for your feedback on the design.

Yes, I do agree that it is too tightly based on the current implementation, especially with regards to two types of timeseries that we have currently.

In terms of reducing the complexity, it again comes down to supporting just a single key as per the initial requirement. This would resolve much of the varying output scenarios that the current implementation has such as the [] vs [0] case.

Motivation for this design:
In the original issue sample code usage [here](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46), I saw that first the data was fetched and then count was called. However, I designed this keeping in mind that the user would not have to explicitly state which timeseries the user wants a count for. 
I noticed that the key could be used to identify which timeseries is required and hence went ahead with taking in a key.

Pending issue:
Now, in this design scenario, there was the case of blank keys passed. 
**How would we decide what timeseries count was to be returned to the user for this case?** 
Hence, I went ahead with returning the total count of the two available timeseries.
This resulted in my output now changing from a single value to two separate values for the two timeseries.

Possible resolution:
- Stick to a single key and return count.
- Need to decide how blank keys are handled AND,
- Which timeseries data to be considered if user wants count of all documents without specifying key?


Also, I am a bit unclear what you mean by this - _"Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."_

Reply:
The design should be driven by what is easy for users to use and generalizable. Not by what is easy to implement.

> Motivation for this design:

Motivation for which design? single key or multi-key?

>  How would we decide what timeseries count was to be returned to the user for this case?

I don't see why this is a problem. What does `find_entries` do if the user doesn't specify a key? Do the same things for counts.

> Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."

If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']` returned `([x, y], [0, 0])` then we could do `np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

Reply:
This is why it is good, when you come to a decision point like this, to document the alternatives, think through the pros and cons, and document your decision instead of just refactoring.

Reply:
> Motivation for which design? single key or multi-key?

Single key initially. 
Approach to fetch timeseries data using keys is valid even in case of multi-key.
I moved to multi-key since I followed how find_entries was handling multiple keys. 

--------

> > How would we decide what timeseries count was to be returned to the user for this case?
> 
> I don't see why this is a problem. What does `find_entries` do if the user doesn't specify a key? Do the same things for counts.

Right, so I have done what `find_entries` is doing - in case of blank keys, fetch the total count of each specific dataset, which itself is done via a helper function `_get_entries_for_timeseries`.
Here, `_get_entries_for_timeseries` (used in `find_entries`) , adds up all counts for keys belonging to a dataset to give a single count.
Ex. ['background/location, 'background/filtered_location'] output would be [x+y]

The difference in my implementation of `_get_entries_counts_for_timeseries` is that I am returning individual counts for each key separately an array element part of combined list.
Ex. ['background/location, 'background/filtered_location'] = [x,y]

This is done for each of the two timeseries and their respective query results are returned in `find_entries`.
Similarly, I am returning the respective counts in `find_entries_count`.

-------

> > Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."
> 
> If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']`returned`([x, y], [0, 0])`then we could do`np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Alright, thank you for the clarification. I do get your point. 
This would be a small change to match up the number of entries in the output.

However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.
For instance, if 2nd list also had non-zero values, say [a,b], corresponding to some analysis_timseries keys, each element inside the 1st list refers to a separate key which will not match with the key from 2nd list.
Would this be appropriate: `x (= background/location ) + a = (analysis/some_key)  = x + a`


-----

> Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

With reference to sample code in original issue [here:](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46)
`total_trips = edb.get_analysis_timeseries_db().count_documents(arguments include userid and keys... )`,

Now, the count can be fetched by:
`ts_user = esta.get_time_series(UUID(user_id))`
`key_list = ['analysis/confirmed_trip']`
`total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]`

Reply:
-----

> Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

With reference to sample code in original issue [here:](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46)
`total_trips = edb.get_analysis_timeseries_db().count_documents(arguments include userid and keys... )`,

Now, the count can be fetched by:
`ts_user = esta.get_time_series(UUID(user_id))`
`key_list = ['analysis/confirmed_trip']`
`total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]`

In case of multiple keys, it would not be easy to access by indexing. So, a better way to return counts would be a dictionary with key-value pair as key_name : count.

Reply:
-------

> > Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."
> 
> If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']`returned`([x, y], [0, 0])`then we could do`np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Alright, thank you for the clarification. I do get your point. 
This would be a small change to match up the number of entries in the output.

However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.
For instance, if 2nd list also had non-zero values, say [a,b], corresponding to some analysis_timseries keys, each element inside the 1st list refers to a separate key which will not match with the key from 2nd list.
Would this be appropriate: `x (= background/location ) + a = (analysis/some_key)  = x + a`

Reply:
> Right, so I have done what find_entries is doing - in case of blank keys, fetch the total count of each specific dataset, which itself is done via a helper function _get_entries_for_timeseries.
> Here, _get_entries_for_timeseries (used in find_entries) , adds up all counts for keys belonging to a dataset to give a single count.

This is the **implementation** of `find_entries`. The **interface** of `find_entries` is that if you give it blank keys, you get a combined list of all entries in the time range. The count is not even exposed outside the implementation. The **interface** of `find_entry_counts` should be the same as the **interface** of `find_entries`, which is that it should return the number of matching entries for that time range.

> This is done for each of the two timeseries and their respective query results are returned in find_entries.

The respective query results are not returned in `find_entries`. A combined query result is returned.
See `itertools.chain(orig_ts_db_result, analysis_ts_db_result)`

> However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.

I fail to see who would be interested in the implementation detail that we have two timeseries collections internally, or why they would need to know which key is stored while accessing the result.

```
ts_user = esta.get_time_series(UUID(user_id))
key_list = ['analysis/confirmed_trip']
total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]
```

and if it were location, we would need:

```
ts_user = esta.get_time_series(UUID(user_id))
key_list = ['background/location']
total_trips = ts_user.find_entries_count(key_list=key_list)[0][0]
```

can you see how that is leaking the implementation outside the interface?